pkg/generate/code/late_initialization.go

```diff
@@ -0,0 +1,246 @@
+	out += fmt.Sprintf("}\n")
+	out += fmt.Sprintf("}\n")
+	return out
+}
+
+func lateInitSlice(responsePath, crPath string, r *model.CRD, respShapeRef *awssdkmodel.ShapeRef, level int) string {
+	// NOTE(muvaf): Late initialization for slices is not a full-featured late-init.
+	// It works only if the slice in CRD is empty. If there is even one element,
+	// we don't late-init since slices do not have unique identifier like map/struct.
+	// We're playing conservative and not rely on index so that we don't risk
+	// overriding desired state with what's observed.
+	// TODO(muvaf): Kubernetes Apply uses comment markers on fields to specify an
+	// identifier (`name` most of the time). If we can do such thing, we can implement
+	// full late-init for slice elements as well, just like maps.
+	out := fmt.Sprintf(
+		"if len(%s) != 0 && len(%s) == 0 {\n", responsePath, crPath,
+	)
+	out += fmt.Sprintf("%s = make([]%s, len(%s))\n", crPath, GetCorrespondingCRDType(respShapeRef.Shape.MemberRef.Shape, r, true), responsePath)
+	out += fmt.Sprintf("for i%d := range %s {\n", level, responsePath)
+	respFieldPath := fmt.Sprintf("%s[i%d]", responsePath, level)
+	crFieldPath := fmt.Sprintf("%s[i%d]", crPath, level)
+	out += lateInit(respFieldPath, crFieldPath, r, &respShapeRef.Shape.MemberRef, level+1)
+	out += fmt.Sprintf("}\n")
+	out += fmt.Sprintf("}\n")
+	return out
+}
+
+func lateInit(responsePath, crPath string, r *model.CRD, str *awssdkmodel.ShapeRef, level int) string {
+	switch str.Shape.Type {
+	case "string":
+		return fmt.Sprintf("%s = awsclients.LateInitializeStringPtr(%s, %s)\n", crPath, crPath, responsePath)
+	case "long", "integer":
+		return fmt.Sprintf("%s = awsclients.LateInitializeInt64Ptr(%s, %s)\n", crPath, crPath, responsePath)
+	case "boolean":
+		return fmt.Sprintf("%s = awsclients.LateInitializeBoolPtr(%s, %s)\n", crPath, crPath, responsePath)
+	case "list":
+		return lateInitSlice(responsePath, crPath, r, str, level)
+	case "structure":
+		return lateInitStruct(responsePath, crPath, r, str, level)
+	case "map":
+		return lateInitMap(responsePath, crPath, r, str, level)
+	default:
+		panic(fmt.Sprintf("unknown shape type %s", str.Shape.Type))
+	}
+}
+
+func GetCorrespondingCRDType(s *awssdkmodel.Shape, r *model.CRD, keepPointer bool) string {
+	switch s.Type {
+	case "string":
+		return "*string"
+	case "long", "integer":
+		return "*int64"
+	case "boolean":
+		return "*bool"
+	case "list":
+		return fmt.Sprintf("[]%s", GetCorrespondingCRDType(s.ValueRef.Shape, r, true))
+	case "map":
+		return fmt.Sprintf("map[string]%s", GetCorrespondingCRDType(s.ValueRef.Shape, r, true))
+	}
+	goType := s.GoTypeWithPkgName()
+	goType = model.ReplacePkgName(goType, r.SDKAPIPackageName(), "svcapitypes", keepPointer)
+	goTypeNoPkg := strings.Split(goType, ".")[1]
+	goPkg := strings.Split(goType, ".")[0]
+	if r.TypeRenames()[goTypeNoPkg] != "" {
+		goTypeNoPkg = r.TypeRenames()[goTypeNoPkg]
+	}
+	return fmt.Sprintf("%s.%s", goPkg, goTypeNoPkg)
+}
```

Comment:
The above seems to be reimplementing some parts of this function:

https://github.com/aws-controllers-k8s/code-generator/blob/41d11f752387871a48a9a88df99b7a751af8a75e/pkg/model/types.go#L28-L74

and some parts of these functions:

https://github.com/aws/aws-sdk-go/blob/27d5e03d86c284f01aabb029636efb92ec3d167e/private/model/api/shape.go#L306-L428

Is that your intention here? If so, you may be able to simplify the code above by either exporting the `pkg/model:cleanGoType` function and/or directly using some of the `aws-sdk-go/private/model/api:Shape` methods.

Reply:
`pkg/model:cleanGoType` seems to calculate the new type name for the struct in the CRD package but here we need something similar to [`varEmptyConstructorK8sType`](https://github.com/aws-controllers-k8s/code-generator/blob/eef7ca5/pkg/generate/code/set_sdk.go#L1094) because the type string is used to write the initialization statement. In fact, it seems like we can reuse the `GetCorrespondingCRDType` function there as well.

> using some of the aws-sdk-go/private/model/api:Shape methods.

I believe except for `strings` operations, it's all methods of the `Shape` object already but seems like we don't need to take every shape type into account as `GoTypeWithPkgName`/`GoType` functions already do that for us. Pushed an update.